Shopify/ruby-lsp
# Add error telemetry and resiliency

### Motivation

If there's a bug in the LSP and a request raises an exception, instead of letting the server process blow up and hit the user with an error dialogue, we should fail gracefully.

Additionally, we can send the error information to the extension's side telemetry, so that the errors can be recorded and addressed.

### Implementation

There are two changes to make error handling better in this PR:
1. We were initially not writing any results that were `nil` to the output pipe. However, this is not correct, since most requests can actually respond with `null`. By not writing a `nil` result back if VS Code is expecting one, we can actually make the editor hang, which is what happened previously with the code actions request. In this PR, we are changing it so that only `false` being returned from a handler prevents writing to the output pipe and `nil` still goes through to avoid having the editor hang waiting for requests
2. We are now rescuing any `StandardError` occurrences, including that information in the telemetry and responding with `nil` for erroring requests. This allows us to fail gracefully while still responding to VS Code, which prevents requests from hanging

### Automated Tests

Let me know if you have ideas on how to test it.

### Manual Tests

1. On `vscode-ruby-lsp`, start the plugin in debug mode using this branch of the `ruby-lsp`
2. Force some request to raise an error. For example, add `raise StandardError, "boom!"` in any handler defined in the `Cli`
3. Make sure to reload the window to restart the server
4. Open some files to trigger requests
5. In the `vscode-ruby-lsp` window, check the debug console logs
6. Verify that errors are logged through telemetry
7. Verify that the LSP server doesn't break and quit
